src/Microsoft.AspNetCore.Razor.Runtime/Runtime/TagHelpers/TagHelperExecutionContext.cs

```diff
@@ -146,12 +146,8 @@ public void AddMinimizedHtmlAttribute(string name)
                 throw new ArgumentNullException(nameof(name));
             }
 
-            EnsureHtmlAttributes();
-            EnsureAllAttributes();
-
             var attribute = new TagHelperAttribute(name);
-            _htmlAttributes.Add(attribute);
-            _allAttributes.Add(attribute);
+            AddHtmlAttribute(attribute);
```

Comment:
Hmm we could cache at runtime. Only downside is creating the dictionary to hold the cached objects; however, it would be a much simpler implementation (than doing the code gen approach).

Are you suggesting doing the cache at runtime or having an extra layer of caching to catch dynamicish bits?